golang/cosmos/x/swingset/genesis.go

```diff
@@ -169,32 +190,72 @@ type swingStoreGenesisEventHandler struct {
 	snapshotHeight uint64
 	swingStore     sdk.KVStore
 	hasher         hash.Hash
+	exportMode     string
 }
 
 func (eventHandler swingStoreGenesisEventHandler) OnExportStarted(height uint64, retrieveSwingStoreExport func() error) error {
 	return retrieveSwingStoreExport()
 }
 
 func (eventHandler swingStoreGenesisEventHandler) OnExportRetrieved(provider keeper.SwingStoreExportProvider) error {
-	if eventHandler.snapshotHeight != provider.BlockHeight {
+	if !((eventHandler.exportMode == keeper.SwingStoreArtifactModeDebug && eventHandler.snapshotHeight == 0) ||
+		eventHandler.snapshotHeight == provider.BlockHeight) {
 		return fmt.Errorf("snapshot block height (%d) doesn't match requested height (%d)", provider.BlockHeight, eventHandler.snapshotHeight)
 	}
 
+	artifactsEnded := false
+
+	var getExportDataReader = func() (agoric.KVEntryReader, error) {
+		exportDataIterator := eventHandler.swingStore.Iterator(nil, nil)
+		kvReader := agoric.NewKVIteratorReader(exportDataIterator)
+		eventHandler.hasher.Reset()
+		encoder := json.NewEncoder(eventHandler.hasher)
+		encoder.SetEscapeHTML(false)
+
+		return agoric.NewKVHookingReader(kvReader, func(entry agoric.KVEntry) error {
+			return encoder.Encode(entry)
+		}, func() error {
+			return nil
+		}), nil
+	}
+
 	artifactsProvider := keeper.SwingStoreExportProvider{
-		GetExportDataReader: func() (agoric.KVEntryReader, error) {
-			exportDataIterator := eventHandler.swingStore.Iterator(nil, nil)
-			kvReader := agoric.NewKVIteratorReader(exportDataIterator)
-			eventHandler.hasher.Reset()
-			encoder := json.NewEncoder(eventHandler.hasher)
-			encoder.SetEscapeHTML(false)
+		GetExportDataReader: getExportDataReader,
+		ReadNextArtifact: func() (types.SwingStoreArtifact, error) {
+			var (
+				artifact          types.SwingStoreArtifact
+				err               error
+				encodedExportData bytes.Buffer
+			)
 
-			return agoric.NewKVHookingReader(kvReader, func(entry agoric.KVEntry) error {
-				return encoder.Encode(entry)
-			}, func() error {
-				return nil
-			}), nil
+			if !artifactsEnded {
+				artifact, err = provider.ReadNextArtifact()
+			} else {
+				return types.SwingStoreArtifact{}, io.EOF
+			}
+
+			if err == io.EOF {
+				artifactsEnded = true
+				if eventHandler.exportMode == keeper.SwingStoreArtifactModeDebug {
+					exportDataReader, _ := getExportDataReader()
+
+					defer exportDataReader.Close()
+
+					err = agoric.EncodeKVEntryReaderToJsonl(
+						exportDataReader,
+						&encodedExportData,
+					)
+					if err == nil {
+						artifact = types.SwingStoreArtifact{
+							Data: encodedExportData.Bytes(),
+							Name: keeper.UntrustedExportDataArtifactName,
+						}
+					}
```

Comment:
It really is a bit unfortunate we have to read, decode, re-encode it all in memory before finally writing it out, but streamlining this is definitely out of scope.